trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/BuchiInterpolantAutomatonBuilder.java

```diff
@@ -0,0 +1,209 @@
+import de.uni_freiburg.informatik.ultimate.lib.modelcheckerutils.smt.predicates.PredicateFactory;
+import de.uni_freiburg.informatik.ultimate.lib.modelcheckerutils.smt.predicates.PredicateUnifier;
+import de.uni_freiburg.informatik.ultimate.lib.smtlibutils.SmtUtils.SimplificationTechnique;
+import de.uni_freiburg.informatik.ultimate.lib.smtlibutils.SmtUtils.XnfConversionTechnique;
+import de.uni_freiburg.informatik.ultimate.lib.tracecheckerutils.singletracecheck.InterpolationTechnique;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.PredicateFactoryForInterpolantAutomata;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.interpolantautomata.transitionappender.NondeterministicInterpolantAutomaton;
+
+/**
+ * @author Matthias Heizmann (heizmann@informatik.uni-freiburg.de)
+ * @author Daniel Dietsch (dietsch@informatik.uni-freiburg.de)
+ * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
+ */
+public class BuchiInterpolantAutomatonBuilder<LETTER extends IIcfgTransition<?>> {
+	private final IUltimateServiceProvider mServices;
+	private final CfgSmtToolkit mCsToolkit;
+	private final SimplificationTechnique mSimplificationTechnique;
+	private final XnfConversionTechnique mXnfConversionTechnique;
+	private final PredicateFactory mPredicateFactory;
+	private final PredicateFactoryForInterpolantAutomata mStateFactoryInterpolAutom;
+
+	public BuchiInterpolantAutomatonBuilder(final IUltimateServiceProvider services, final CfgSmtToolkit csToolkit,
+			final SimplificationTechnique simplificationTechnique, final XnfConversionTechnique xnfConversionTechnique,
+			final PredicateFactory predicateFactory,
+			final PredicateFactoryForInterpolantAutomata stateFactoryInterpolAutom) {
+		mServices = services;
+		mCsToolkit = csToolkit;
+		mSimplificationTechnique = simplificationTechnique;
+		mXnfConversionTechnique = xnfConversionTechnique;
+		mPredicateFactory = predicateFactory;
+		mStateFactoryInterpolAutom = stateFactoryInterpolAutom;
+	}
+
+	public NestedWordAutomaton<LETTER, IPredicate> constructInterpolantAutomaton(final IPredicate precondition,
+			final NestedLassoRun<LETTER, ?> counterexample, final IPredicate[] stemInterpolants, final IPredicate honda,
+			final IPredicate[] loopInterpolants, final VpAlphabet<LETTER> vpAlphabet,
+			final IEmptyStackStateFactory<IPredicate> emptyStateFactory) {
+		final NestedWord<LETTER> stem = counterexample.getStem().getWord();
+		final NestedWord<LETTER> loop = counterexample.getLoop().getWord();
+		final NestedWordAutomaton<LETTER, IPredicate> result =
+				new NestedWordAutomaton<>(new AutomataLibraryServices(mServices), vpAlphabet, emptyStateFactory);
+		if (stem.length() == 0) {
+			result.addState(true, true, honda);
+		} else {
+			result.addState(true, false, precondition);
+			for (int i = 0; i < stemInterpolants.length; i++) {
+				addState(stemInterpolants[i], result);
```

Comment:
why do you call `addState` sometimes directly on the NWA and sometimes via the `addState` wrapper?

Reply:
Not sure, this is just some old unchanged code that was moved to an extra class.